packages/next/src/server/lib/patch-fetch.ts

```diff
@@ -348,45 +340,54 @@ export function patchFetch({
           (hasUnCacheableHeader || isUnCacheableMethod) &&
           staticGenerationStore.revalidate === 0
 
-        if (isForceNoStore) {
-          cacheReason = 'fetchCache = force-no-store'
-        }
-
-        if (isOnlyNoStore) {
-          if (
-            _cache === 'force-cache' ||
-            (typeof revalidate !== 'undefined' &&
-              (revalidate === false || revalidate > 0))
-          ) {
-            throw new Error(
-              `cache: 'force-cache' used on fetch for ${fetchUrl} with 'export const fetchCache = 'only-no-store'`
-            )
+        switch (fetchCacheMode) {
+          case 'force-no-store': {
+            cacheReason = 'fetchCache = force-no-store'
+            break
           }
-          cacheReason = 'fetchCache = only-no-store'
-        }
-
-        if (isOnlyCache && _cache === 'no-store') {
-          throw new Error(
-            `cache: 'no-store' used on fetch for ${fetchUrl} with 'export const fetchCache = 'only-cache'`
-          )
-        }
-
-        if (
-          isForceCache &&
-          (typeof curRevalidate === 'undefined' || curRevalidate === 0)
-        ) {
-          cacheReason = 'fetchCache = force-cache'
-          revalidate = false
+          case 'only-no-store': {
+            if (
+              _cache === 'force-cache' ||
+              (typeof revalidate !== 'undefined' &&
+                (revalidate === false || revalidate > 0))
+            ) {
+              throw new Error(
+                `cache: 'force-cache' used on fetch for ${fetchUrl} with 'export const fetchCache = 'only-no-store'`
+              )
+            }
+            cacheReason = 'fetchCache = only-no-store'
+            break
+          }
+          case 'only-cache': {
+            if (_cache === 'no-store') {
+              throw new Error(
+                `cache: 'no-store' used on fetch for ${fetchUrl} with 'export const fetchCache = 'only-cache'`
+              )
+            }
+            break
+          }
+          case 'force-cache': {
+            if (typeof curRevalidate === 'undefined' || curRevalidate === 0) {
+              cacheReason = 'fetchCache = force-cache'
+              revalidate = false
+            }
+            break
+          }
+          default:
+          // sometimes we won't match the above cases. the reason we don't move
+          // everything to this switch is the use of autoNoCache which is not a fetchCacheMode
+          // I suspect this could be unified with fetchCacheMode however in which case we could
+          // simplify the switch case and ensure we have an exhaustive switch handling all modes
```

Comment:
To avoid bugs related to ordering of `switch/case`, I'd add a `break` for the `default` to ensure that if cases are added afterwards it won't fall through. ([See example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAbgQwDYFcCmAuG0BOBLMAcxgF4YAiKNacgKFogHc8pgALGACkVTQEoYAb1oxRMYAghoKVchhFjFoSCCRoAdEhCFOlNOT4LFogEY40CANZGYAEzQAzBCiRR5xsQHpPNiVJnUcjZK4BCqGlo6ehAGwWJmFtYAvkA))

Reply:
checked and lint will yell at you if you put a new case below the default